Implement decoupled state commitments and storage

roysc opened the pull request:
Introduces a new `CommitKVStore` to implement [ADR-040 proposal](https://github.com/cosmos/cosmos-sdk/pull/8430) of decoupled commitment/data concerns.

Adapted from `store/ll-smt` branch (see https://github.com/cosmos/cosmos-sdk/pull/8507).

<!-- < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < < ☺
v                               ✰  Thanks for creating a PR! ✰
v    Before smashing the submit button please review the checkboxes.
v    If a checkbox is n/a - please still include it but + a little note why
☺ > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > > >  -->

## Description

Introduces the `decoupled.Store` type composed of:
* A tm-db interface for mapping state storage (SS) (as `key → value`)
* A "state commitment" (SC) store for computing Merkle root hashes (storing `hash(key) → hash(key, value)`)
  * This is an `iavl.Store` held as a `CommitKVStore` reference, which will later be refactored to use an SMT store
* A tm-db interface for an inverted index, mapping SC values back to SS keys (`hash(key, value) → key`)

closes: #1 

---

Before we can merge this PR, please make sure that all the following items have been
checked off. If any of the checklist items are not applicable, please leave them but
write a little note why.

- [ ] Targeted PR against correct branch (see [CONTRIBUTING.md](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#pr-targeting))
- [x] Linked to Github issue with discussion and accepted design OR link to spec that describes this work.
- [ ] Code follows the [module structure standards](https://github.com/cosmos/cosmos-sdk/blob/master/docs/building-modules/structure.md). N/A, no modules
- [ ] Wrote unit and integration [tests](https://github.com/cosmos/cosmos-sdk/blob/master/CONTRIBUTING.md#testing)
- [ ] Updated relevant documentation (`docs/`) or specification (`x/<module>/spec/`)
- [ ] Added relevant `godoc` [comments](https://blog.golang.org/godoc-documenting-go-code).
- [ ] Added a relevant changelog entry to the `Unreleased` section in `CHANGELOG.md`
- [ ] Re-reviewed `Files changed` in the Github PR explorer
- [ ] Review `Codecov Report` in the comment section below once CI passes

roysc committed:
[WIP] skeleton of decoupled.Store

Outline of new store following ADR-040 pattern of decoupled
commitment/data concerns

Largely adapted from store/ll-smt branch

roysc committed:
[WIP] decoupled store; basic tests

i-norden commented on store/decoupled/store.go:
```diff
@@ -0,0 +1,161 @@
+	if err != nil {
+		panic(err.Error())
+	}
+	s.sc.Set(key, kvHash[:]) // TODO: key or hash(key)?
```
Need an `err :=` here. But actually I don't think we need this `Set`, the `key, value` and `hash(key, value), key` will be sufficient

roysc replied:
This is where the state commitment mapping is set; [`Set()`](https://github.com/vulcanize/cosmos-sdk/blob/82dc0081fdc06fa6cf8bd6a0d87987d27ac7b1ca/store/iavl/store.go#L169) has no return value

i-norden replied:
Got it, we don't need the `err` check below then.

i-norden commented on store/decoupled/store.go:
```diff
@@ -0,0 +1,161 @@
+	// Direct KV mapping (SS)
+	data dbm.DB
+	// Inverted index of SC values to SS keys
+	inv dbm.DB
```
Does the second bucket need to move through a separate tm-db interface?

roysc replied:
I'm using `PrefixDB`s to separate the two into different namespaces. Collisions are unlikely since the inverted index uses a hash as key, but I think it's still desired.

i-norden reviewed (commented):
Looking good. In addition to the comments, as you know we need to replace tm-db with tm-dbv2 and remove IAVL usage. In addition to that the next step is to flesh out how the decoupling will be handled at the multistore level. I think what we want to do is have the multistore map StoreKeys to a Store tuple (SCKVStore, SSKVStore).

roysc committed:
[wip] rf decoupled store into multistore

roysc committed:
fix test

i-norden commented on store/decoupled/store.go:
```diff
@@ -0,0 +1,212 @@
+	if err != nil {
+		panic(err.Error())
+	}
+	err = s.inv.Set(kvHash[:], key)
```
We're deviating pretty hard from the simple behavior expected from a KVStore. A KVStore `Set` is expected to `Set` that key-value, it isn't expected to generate and `Set` two other kv pairs. I think this logic makes more sense at a higher level.

roysc replied:
Per standup discussion, let's run this by Cosmos to see if this is the intended semantics, or where it's more appropriate to put this logic.

i-norden commented on store/decoupled/store.go:
```diff
@@ -0,0 +1,212 @@
+}
+
+func (s *Store) Iterator(start, end []byte) types.Iterator {
+	iter, err := s.data.Iterator(start, end)
```
I think methods such as this one also highlight the issue with wrapping 3 distinct KVStores with a single KVStore- when we ask for an Iterator on this wrapping KVStore we are only getting the iterator for one of the underlying KVStores.

i-norden commented on store/decoupled/store.go:
```diff
@@ -0,0 +1,212 @@
+}
+
+func (s *Store) LastCommitID() types.CommitID {
+	return s.sc.LastCommitID()
```
This also highlights the issue. Which underlying commit ID are we asking for?

roysc committed:
test case for bucket independence

roysc committed:
test basic merkle hash

i-norden commented on store/decoupled/store.go:
```diff
@@ -0,0 +1,212 @@
+	if err != nil {
+		panic(err.Error())
+	}
+	kvHash := sha256.Sum256(append(key, value...))
```
Instead of looking up the value and hashing it and the key to reconstruct `hash(key, value)` we could lookup the `hash(key, value)` from the SC before deleting it. This appears preferable since both approaches require a DB lookup but this approach doesn't require a hashing step. What do you think?

roysc replied:
That does indeed make more sense

roysc committed:
no need to re-hash when deleting from index

roysc committed:
fixup

tzdybal committed:
Add SMT store type (#8507)

* Initial SMT store type

SMT (Sparse Merkle Tree) is intended to replace IAVL. New type
implements same interfaces as iavl.Store.

* Add iteration support to SMT

Sparse Merkle Tree does not support iteration over keys in order.
To provide drop-in replacement for IAVL, Iterator and ReverseIterator
has to be implemented.
SMT Store implementation use the underlying KV store to:
 - maintain a list of keys (under a prefix)
 - iterate over a keys
Values are stored only in SMT.

* Migrate to smt v0.1.1

* Extra test for SMT iterator

* CommitStore implementation for SMT store

* Use interface instead of concrete type

* Add telemetry to SMT store

* SMT: version->root mapping, cleanup

* SMT proofs - initial code

* Tests for SMT store ProofOp implementation

* Fix linter errors

* Use simple 1 byte KV-store prefixes

* Improve assertions in tests

* Use mutex properly

* Store data in ADR-040-compatible way

SMT stores:
 * key -> hash(key, value)

KV store stores:
 * key->value in "bucket 1",
 * hash(key, value) -> key in "bucket 2".

robert-zaremba committed:
Merge branch 'master' into store/ll-smt

roysc committed:
decoupled store query, versioning

roysc committed:
replace tm-db mod for interface additions, refactor

roysc committed:
go mod tidy

roysc commented:
This should be a MVP state store for #1.
Requires changes from the corresponding branch on tm-db: https://github.com/roysc/tm-db/tree/implement-adr-40 - (which just patches in minimal versioning support)

roysc committed:
Refactor for new DB interface

- Move tm-db code in from external repo
- Refactor store/ and types/ packages
- leaves out IAVL and multistore code pending confirmed deprecation
- some hacks to glue existing tmdb usage and get compile/tests working

roysc committed:
[wip] decoupled.Store

- Update to new db interface
- Use smt as SC structure
- Refactor smt to BasicKVStore

roysc committed:
Update modules

roysc commented:
Moving this work to a PR on the main repo: https://github.com/cosmos/cosmos-sdk/pull/9485
